crypto/evp/keymgmt_meth.c

```diff
@@ -47,154 +47,99 @@ static void *keymgmt_from_dispatch(int name_id,
+            if (keymgmt->validatedomparams == NULL)
+                keymgmt->validatedomparams =
+                    OSSL_get_OP_keymgmt_validate_domparams(fns);
             break;
         case OSSL_FUNC_KEYMGMT_VALIDATE_PUBLIC:
-            if (keymgmt->validatepublic != NULL)
-                break;
-            keymgmt->validatepublic =
-                OSSL_get_OP_keymgmt_validate_public(fns);
+            if (keymgmt->validatepublic == NULL)
+                keymgmt->validatepublic =
+                    OSSL_get_OP_keymgmt_validate_public(fns);
             break;
         case OSSL_FUNC_KEYMGMT_VALIDATE_PRIVATE:
-            if (keymgmt->validateprivate != NULL)
-                break;
-            keymgmt->validateprivate =
-                OSSL_get_OP_keymgmt_validate_private(fns);
+            if (keymgmt->validateprivate == NULL)
+                keymgmt->validateprivate =
+                    OSSL_get_OP_keymgmt_validate_private(fns);
             break;
         case OSSL_FUNC_KEYMGMT_VALIDATE_PAIRWISE:
-            if (keymgmt->validatepairwise != NULL)
-                break;
-            keymgmt->validatepairwise =
-                OSSL_get_OP_keymgmt_validate_pairwise(fns);
+            if (keymgmt->validatepairwise == NULL)
+                keymgmt->validatepairwise =
+                    OSSL_get_OP_keymgmt_validate_pairwise(fns);
             break;
         }
     }
     /*
      * Try to check that the method is sensible.
+     * At least one constructor and the destructor are MANDATORY
+     * The functions has_public_key an has_private_key are MANDATORY
      * It makes no sense being able to free stuff if you can't create it.
      * It makes no sense providing OSSL_PARAM descriptors for import and
      * export if you can't import or export.
      */
-    if ((keymgmt->freedomparams != NULL
-         && (keymgmt->importdomparams == NULL
-             && keymgmt->gendomparams == NULL))
-        || (keymgmt->freekey != NULL
-            && (keymgmt->importkey == NULL
-                && keymgmt->genkey == NULL
-                && keymgmt->loadkey == NULL))
-        || (keymgmt->importdomparam_types != NULL
-            && keymgmt->importdomparams == NULL)
-        || (keymgmt->exportdomparam_types != NULL
-            && keymgmt->exportdomparams == NULL)
-        || (keymgmt->gettable_domparam_params != NULL
-            && keymgmt->get_domparam_params == NULL)
-        || (keymgmt->importkey_types != NULL
-            && keymgmt->importkey == NULL)
-        || (keymgmt->exportkey_types != NULL
-            && keymgmt->exportkey == NULL)
-        || (keymgmt->gettable_key_params != NULL
-            && keymgmt->get_key_params == NULL)) {
+    if (keymgmt->free == NULL
+        || keymgmt->new == NULL
+        || keymgmt->has_public_key == NULL
+        || keymgmt->has_private_key == NULL
+        || (keymgmt->gettable_params != NULL
+            && keymgmt->get_params == NULL)
```

Comment:
This hasn't actually been made a rule yet...  I'm not 100% convinced yet

Reply:
I looked over all other method constructors, and apart from not being perfectly consistent (some constructors do not check this at all, but all others do check that either both are present or none of them are), it appears that this interface is the only one only checking one way.

So ok, I'll bow down and implement one of the schemes instead of this half-and-half scheme.